src/Agent.Worker/Release/Artifacts/BuildArtifact.cs

```diff
@@ -0,0 +1,242 @@
+            ArgUtil.NotNull(vssEndpoint, nameof(vssEndpoint));
+            ArgUtil.NotNull(vssEndpoint.Url, nameof(vssEndpoint.Url));
+
+            // TODO Get this value from settings
+            string parallelDownloadLimit = "8";
+
+            var artifactDetails = JsonConvert.DeserializeObject<Dictionary<string, string>>(agentArtifactDefinition.Details);
+            VssCredentials vssCredentials = ApiUtil.GetVssCredential(vssEndpoint);
+            var tfsUrl = context.Variables.Get(WellKnownDistributedTaskVariables.TFCollectionUrl);
+
+            Guid projectId = context.Variables.System_TeamProjectId ?? Guid.Empty;
+            ArgUtil.NotEmpty(projectId, nameof(projectId));
+
+            string relativePath;
+            string accessToken;
+            vssEndpoint.Authorization.Parameters.TryGetValue(EndpointAuthorizationParameters.AccessToken, out accessToken);
+
+            if (artifactDetails.TryGetValue("RelativePath", out relativePath))
+            {
+                return new BuildArtifactDetails
+                {
+                    Credentials = vssCredentials,
+                    RelativePath = artifactDetails["RelativePath"],
+                    Project = projectId.ToString(),
+                    TfsUrl = new Uri(tfsUrl),
+                    AccessToken = accessToken,
+                    ParallelDownloadLimit =
+                                   Convert.ToInt32(
+                                       parallelDownloadLimit,
+                                       CultureInfo.InvariantCulture)
+                };
+            }
+            else
+            {
+                throw new InvalidOperationException(StringUtil.Loc("RMArtifactDetailsIncomplete"));
+            }
+        }
+
+        private bool Match(ServerBuildArtifact buildArtifact, ArtifactDefinition artifactDefinition)
+        {
+            // If this is older , then dont force the name checks. 
+            if (string.Equals(artifactDefinition.Name, AllArtifacts, StringComparison.OrdinalIgnoreCase))
+            {
+                return true;
+            }
+
+            if (string.Equals(artifactDefinition.Name, buildArtifact.Name, StringComparison.OrdinalIgnoreCase))
+            {
+                return true;
+            }
+
+            return false;
+        }
+
+        private async Task DownloadArtifact(
+            IExecutionContext executionContext,
+            ServerBuildArtifact buildArtifact,
+            ArtifactDefinition artifactDefinition,
+            string localFolderPath,
+            BuildHttpClient buildClient,
+            XamlBuildHttpClient xamlBuildClient,
+            DefinitionType definitionType,
+            int buildId)
+        {
+            var downloadFolderPath = Path.Combine(localFolderPath, buildArtifact.Name);
+            var buildArtifactDetails = artifactDefinition.Details as BuildArtifactDetails;
```

Comment:
Not required. The caller doing a foreach and sending buildArtifact down. Also the other parameters already validated in the caller itself.